// containers/product/index.js
import React, {Component} from 'react'
import {connect} from 'react-redux'
import R from 'ramda'
import {Link} from 'react-router'

import {fetchProductById,addProductToBasket} from 'actions'
import {getProductById} from 'selectors'
import BasketCart from 'components/basketCart'

class Product extends Component {

    componentDidMount() {
        this.props.fetchProductById(this.props.params.id)
    }

    renderFields() {
        const {product} = this.props;
        const columnFields = R.compose(
            R.toPairs,
            R.pick([
                'cpu',
                'camera',
                'size',
                'weight',
                'display',
                'battery',
                'memory'
            ])
        )(product);


        return columnFields.map(([key,value]) => (
            <div className="column" key={key}>
                <div className="ab-details-title">
                    <p className="text-uppercase">{key}</p>
                </div>
                <div className="ab-details-info">
                    <p>{value}</p>
                </div>
            </div>
        ))
    }

    renderContent() {
        const {product} = this.props;

        return(
            <div className="thumbnail">
                <div className="row">
                    <div className="col-md-6">
                        <img
                            src={product.image}
                            alt={product.name}
                            className="img-thumbnail"
                        />
                    </div>
                    <div className="col-md-6">
                        {this.renderFields()}
                    </div>
                </div>
                <div className="caption-full">
                    <h4 className="pull-right">${product.price}</h4>
                    <h4>{product.name}</h4>
                    <p>{product.description}</p>
                </div>
            </div>
        )
    }

    renderSidebar() {
        const {product, addProductToBasket} = this.props;
        return(
            <div>
                <p className="lead">Quick shop</p>
                <BasketCart />
                <div className="from-group">
                    <h1>{product.name}</h1>
                    <h2>${product.price}</h2>
                </div>
                <button
                    type="button"
                    className="btn btn-success btn-block"
                    onClick={()=> addProductToBasket(product.id)}
                >
                    Add to cart
                </button>
                <Link
                    to="/"
                    className="btn btn-info btn-block"
                >
                    Back to store
                </Link>
            </div>
        )
    }

    render() {

        const {product} = this.props;
        return (
            <div className="view-container">
                <div className="container">
                    <div className="row">
                        <div className="col-md-9">
                            {product && this.renderContent()}
                        </div>
                        <div className="col-md-3">
                            {product && this.renderSidebar()}
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

const mapStateToProps = state => ({
   product: getProductById(state, state.product.id)
});

const mapDispatchToProps = {
    fetchProductById,
    addProductToBasket
};

export default connect(mapStateToProps, mapDispatchToProps)(Product)
